src/cdk/testing/harness-environment.ts

```diff
@@ -147,43 +130,179 @@ export abstract class HarnessEnvironment<E> implements HarnessLoader, LocatorFac
+      Promise<(LocatorFnResult<T>)[]> {
+    return Promise.all(rawElements.map(rawElement => {
+      const testElement = this.createTestElement(rawElement);
+      return Promise.all(
+          // For each query, get `null` if it doesn't match, or a `TestElement` or
+          // `ComponentHarness` as appropriate if it does match. This gives us everything that
+          // matches the current raw element, but it may contain duplicate entries (e.g. multiple
+          // `TestElement` or multiple `ComponentHarness` of the same type.
+          allQueries.map(query => this._getQueryResultForElement(query, rawElement, testElement))
+      ).then(allResultsForElement => _removeDuplicateQueryResults(allResultsForElement));
+    })).then(perElementMatches => ([] as any).concat(...perElementMatches));
   }
 
-  private async _assertElementFound(selector: string): Promise<E> {
-    const element = (await this.getAllRawElements(selector))[0];
-    if (!element) {
-      throw Error(`Expected to find element matching selector: "${selector}", but none was found`);
+  /**
+   * Check whether the given query matches the given element, if it does return the matched
+   * `TestElement` or `ComponentHarness`, if it does not, return null.
+   */
+  private async _getQueryResultForElement<T extends ComponentHarness>(
+      query: string | HarnessPredicate<T>, rawElement: E, testElement: TestElement):
+      Promise<T | TestElement | null> {
+    if (typeof query === 'string') {
+      return testElement.matchesSelector(query)
+          .then(selectorMatches => selectorMatches ? testElement : null);
     }
-    return element;
+    return testElement.matchesSelector(query.getSelector()).then(selectorMatches => {
+      if (selectorMatches) {
+        const harness = this.createComponentHarness(query.harnessType, rawElement);
+        return query.evaluate(harness).then(predMatches => predMatches ? harness : null);
+      }
+      return null;
+    });
   }
+}
 
-  private async _assertHarnessFound<T extends ComponentHarness>(
-      harnessType: ComponentHarnessConstructor<T> | HarnessPredicate<T>): Promise<T> {
-    const harness = (await this._getAllHarnesses(harnessType))[0];
-    if (!harness) {
-      throw _getErrorForMissingHarness(harnessType);
+/**
+ * Parses a list of queries in the format accepted by the `locatorFor*` methods into an easier to
+ * work with format.
+ */
+function _parseQueries<T extends (HarnessQuery<any> | string)[]>(queries: T):
+    ParsedQueries<LocatorFnResult<T> & ComponentHarness> {
+  const allQueries = [];
+  const harnessQueries = [];
+  const elementQueries = [];
+  const harnessTypes =
+      new Set<ComponentHarnessConstructor<LocatorFnResult<T> & ComponentHarness>>();
+
+  for (const query of queries) {
+    if (typeof query === 'string') {
+      allQueries.push(query);
+      elementQueries.push(query);
+    } else {
+      const pred =  query instanceof HarnessPredicate ? query : new HarnessPredicate(query, {});
```

Comment:
Some weird spacing here. Also maybe call it `predicate` since it's not much longer than `pred`.